packages/ui-lib/src/Form/atoms/Switch.tsx

```diff
@@ -191,53 +191,64 @@ interface IProps {
-  const [ justUpdated, setJustUpdated ] = useState<boolean>(false);
-  const [ innerSize, setInnerSize ] = useState<number>(0);
-
+  const [position, setPosition] = useState<SwitchPosition>(isControlled ? (checked ? SwitchPosition.On : SwitchPosition.Off) : (defaultChecked ? SwitchPosition.On : SwitchPosition.Off));
+  const [activeTheming, setActiveTheming] = useState<string>(isControlled ? (checked ? rightTheme : leftTheme) : (defaultChecked ? rightTheme : leftTheme));
+  const [switchState, setSwitchState] = useState<TypeSwitchState>('default');
+  const [justUpdated, setJustUpdated] = useState<boolean>(false);
+  const [innerSize, setInnerSize] = useState<number>(0);
 
   useEffect(() => {
-    if (isDefaultPropPassed) {
-            setPosition(defaultChecked ? SwitchPosition.On : SwitchPosition.Off);
-          } else {
-            setPosition(inputRef.current?.checked ? SwitchPosition.On : SwitchPosition.Off);
-          }
-  }, [checked, defaultChecked, isDefaultPropPassed]); 
-
-    const positionSwitch = useCallback(() => {
-    if (!isDefaultPropPassed) {
-      setPosition(inputRef.current?.checked ? SwitchPosition.On : SwitchPosition.Off);
-      setActiveTheming(inputRef.current?.checked ? rightTheme : leftTheme);
-    } else {
-    if(inputRef.current?.defaultChecked){
-     setPosition(SwitchPosition.On);
-     setActiveTheming(rightTheme);
-   } else if(!inputRef.current?.defaultChecked){
-     setPosition(SwitchPosition.Off);
-     setActiveTheming(leftTheme);
-   }
-  }
-  }, [setPosition, setActiveTheming, leftTheme, rightTheme, isDefaultPropPassed]); 
+    if (isControlled) {
+      setPosition(checked ? SwitchPosition.On : SwitchPosition.Off);
+      setActiveTheming(checked ? rightTheme : leftTheme);
+    }
+    else{
+      if (!isControlled && inputRef.current) {
+        inputRef.current.checked = internalChecked;
+        setPosition(internalChecked ? SwitchPosition.On : SwitchPosition.Off);
+        setActiveTheming(internalChecked ? rightTheme : leftTheme);
+      }
+    }
+  }, [checked, isControlled, leftTheme, rightTheme, internalChecked]);
 
+  const positionSwitch = useCallback(() => {
+    if (isControlled) {
+      setPosition(checked ? SwitchPosition.On : SwitchPosition.Off);
+      setActiveTheming(checked ? rightTheme : leftTheme);
+    } else {
+      setPosition(internalChecked ? SwitchPosition.On : SwitchPosition.Off);
+      setActiveTheming(internalChecked ? rightTheme : leftTheme);
+    }
+  }, [checked, isControlled, leftTheme, rightTheme, internalChecked]);
 
   const customOnChange = () => {
+    if (state === 'locked' || state === 'disabled') return;
+    
+    if (isControlled) {
+      onChangeCallback?.(!checked);
+    } else {
+      const newChecked = !internalChecked;
+      setInternalChecked(newChecked);
+      onChangeCallback?.(newChecked);
+    }
     positionSwitch();
-    setJustUpdated(true);
-    if(onChangeCallback){ onChangeCallback(inputRef.current?.checked || false); }
   };
```

Comment:
I think this function needs to be wrapped in a "useCallback" to be aware of the internalCheck state change.